component_test.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tfaddr
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	svchost "github.com/hashicorp/terraform-svchost"
+)
+
+func TestParseComponentSource_simple(t *testing.T) {
+	tests := map[string]struct {
+		input   string
+		want    Component
+		wantErr string
+	}{
+		"main registry implied": {
+			input: "hashicorp/k8-cluster",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("registry.terraform.io"),
+					Namespace: "hashicorp",
+					Name:      "k8-cluster",
+				},
+				Subdir: "",
+			},
+		},
+		"main registry implied, subdir": {
+			input: "hashicorp/k8-cluster//path/to/dir",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("registry.terraform.io"),
+					Namespace: "hashicorp",
+					Name:      "k8-cluster",
+				},
+				Subdir: "path/to/dir",
+			},
+		},
+		"custom registry": {
+			input: "terraform.registry.io/hashicorp/k8-cluster",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("terraform.registry.io"),
+					Namespace: "hashicorp",
+					Name:      "k8-cluster",
+				},
+				Subdir: "",
+			},
+		},
+		"custom registry, subdir": {
+			input: "app.terraform.io/hashicorp/k8-cluster//examples/foo",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("app.terraform.io"),
+					Namespace: "hashicorp",
+					Name:      "k8-cluster",
+				},
+				Subdir: "examples/foo",
+			},
+		},
+		"private registry": {
+			input: "example.com/awesomecorp/network",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("example.com"),
+					Namespace: "awesomecorp",
+					Name:      "network",
+				},
+				Subdir: "",
+			},
+		},
+		"private registry, subdir": {
+			input: "example.com/awesomecorp/network//configs",
+			want: Component{
+				Package: ComponentPackage{
+					Host:      svchost.Hostname("example.com"),
+					Namespace: "awesomecorp",
+					Name:      "network",
+				},
+				Subdir: "configs",
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			addr, err := ParseComponentSource(test.input)
+
+			if test.wantErr != "" {
+				switch {
+				case err == nil:
+					t.Errorf("unexpected success\nwant error: %s", test.wantErr)
+				case err.Error() != test.wantErr:
+					t.Errorf("wrong error messages\ngot:  %s\nwant: %s", err.Error(), test.wantErr)
+				}
+				return
+			}
+
+			if err != nil {
+				t.Fatalf("unexpected error: %s", err.Error())
+			}
+
+			if diff := cmp.Diff(addr, test.want); diff != "" {
+				t.Errorf("wrong result\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestParseComponentSource(t *testing.T) {
```

Comment:
Test cases look great! Not required, but wonder if we should include any additional edge cases. Such as including a protocol or query parameters.

Reply:
Added two edge case tests.